python 3 compatibility

dirk-thomas opened the pull request.

tfoote commented:
+1

dirk-thomas committed:
python 3 compatibility

mirzashah committed:
Additional Python 3 compatibility fixes

mirzashah committed:
Fixes for roslaunch for Python 3 compatiblity

mirzashah committed:
Python 3 compatability

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3 with Dirk's feedback

dirk-thomas committed:
Merge pull request #1 from mirzashah/hydro-devel

Python 3 Compatability

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
fix python 2 regressions, still need different python 3 handling

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

mirzashah committed:
Python 3

mirzashah committed:
Python 3

dirk-thomas committed:
python 3 compatibility

mirzashah committed:
Python 3

mirzashah committed:
Merge branch 'hydro-devel' of github.com:dirk-thomas/ros_comm into hydro-devel

mirzashah committed:
Python 3

dirk-thomas committed:
Merge pull request #2 from mirzashah/hydro-devel

Python 3

dirk-thomas committed:
pytho 3 compatibility

dirk-thomas committed:
Merge branch 'hydro-devel' of github.com:dirk-thomas/ros_comm into hydro-devel

dirk-thomas committed:
make order of header fields deterministic

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
rosbag fix indentation of bmr files

dirk-thomas committed:
rosbag python 3 compatibility

dirk-thomas committed:
Merge pull request #3 from dirk-thomas/test

rosbag python 3 compatibility

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Python 3

mirzashah committed:
Merge branch 'hydro-devel' of github.com:mirzashah/ros_comm into hydro-devel

Conflicts:
	tools/rosgraph/test/test_network.py

mirzashah committed:
Python 3

dirk-thomas committed:
Merge pull request #4 from mirzashah/hydro-devel

Python 3

dirk-thomas committed:
python 2 regressions

wjwwood commented on test/test_rosbag/test/connection_count.py:
```diff
@@ -53,7 +53,7 @@ def test_connection_count(self):
 
     conns = False
-    for l in out.split('\n'):
+    for l in out.decode().split('\n'):
```
I'm not sure this will work for Python3 and Python2?

wjwwood commented on test/test_rosbag/test/topic_count.py:
```diff
@@ -55,7 +55,7 @@ def test_topic_count(self):
     out,err = p.communicate()
     topic_count = 0
-    for l in out.split('\n'):
+    for l in out.decode().split('\n'):
```
Same thing here, will `out` have a `.decode()` in Python2?

wjwwood replied:
Nvm, it seems to.

wjwwood commented on test/test_rospy/test/unit/test_rospy_rostime.py:
```diff
@@ -322,11 +328,12 @@ def test_sleep(self):
+        sleepthread = threading.Thread(target=sleeper, args=())
         self.failIf(test_sleep_done)
-        thread.start_new_thread(sleeper, ())
+        sleepthread.start()
```
This should probably also have a `sleepthread.setDaemon(True)` if you want it to be a one-to-one mapping to the `thread.start_new_thread`, see: https://github.com/ros/ros_comm/pull/285#issuecomment-24839576

wjwwood commented on test/test_rospy/test/unit/test_rospy_rostime.py:
```diff
@@ -336,8 +343,9 @@ def test_sleep(self):
-        thread.start_new_thread(duration_sleeper, ())
+        self.failIf(test_duration_sleep_done)      
+        dursleepthread = threading.Thread(target=duration_sleeper, args=())          
+        dursleepthread.start()
```
Same thing here.

wjwwood commented on test/test_rospy/test/unit/test_rospy_rostime.py:
```diff
@@ -348,7 +356,8 @@ def test_sleep(self):
         self.failIf(test_backwards_sleep_done)
-        thread.start_new_thread(backwards_sleeper, ())
+        backsleepthread = threading.Thread(target=backwards_sleeper, args=())          
+        backsleepthread.start()
```
And here.

wjwwood commented on tools/rosbag/src/rosbag/bag.py:
```diff
@@ -892,11 +892,20 @@ def _open(self, f, mode, allow_unindexed):
 
+    def _is_file(self, f):
+        try:
+            return isinstance(f, file) #Python 2
```
According to my editor this comment should have two spaces before the `#` and one after, I believe this is PEP8.

wjwwood commented on tools/rosbag/src/rosbag/bag.py:
```diff
@@ -892,11 +892,20 @@ def _open(self, f, mode, allow_unindexed):
+            return isinstance(f, file) #Python 2
+        except NameError:
+            import io
+            return isinstance(f, io.IOBase) #Python 3...this will return false in Python 2 always
```
Same as above.

wjwwood commented on tools/rosbag/src/rosbag/migration.py:
```diff
@@ -514,10 +515,10 @@ def __init__(self, input_rule_files=[], plugins=True):
             try:
                 scratch_locals = {'MessageUpdateRule':MessageUpdateRule}
-                execfile(r,scratch_locals)
+                exec(open(r).read(), scratch_locals)
```
Ideally this would be:

``` python
with open(r, 'r') as f:
    exec(f.read(), scratch_locals)
```

wjwwood commented on tools/rosgraph/test/test_network.py:
```diff
@@ -56,37 +56,47 @@ def test_encode_ros_handshake_header(self):
+        #the fields are not in the same order as the hand encoded one (variable 'encoded').
+        #The fix for this test is not obvious to me, so please suggest an approach when 
+        #you see this comment. One approach could be to force the encode function to sort the keys alphabetically.
+        #-Mirza        
```
Are we going to address this before merging?

wjwwood commented on tools/rosmaster/test/test_rosmaster_paramserver.py:
```diff
@@ -168,10 +174,21 @@ def test_subscribe_param_tree(self):
-        self.assertEquals([([('ptnode2', 'http://ptnode2:2')], '/gains/p/', 'P3'), \
+
+        #Dirk - This assertion fails, self.last_update is correct but the list is out
+        #of order. I don't think that makes it invalid. - Mirza
```
Another point to be addressed.

wjwwood commented on tools/rosmaster/test/test_rosmaster_registrations.py:
```diff
@@ -272,12 +272,14 @@ def test_Registrations(self):
             self.assertEquals(t, r.type)
-            self.assert_(not r) #test nonzero
+            #Dirk - Before the assert was assert_(not r). That always returns false
+            #and fails for a valid r object. Not sure how this was passing before. -Mirza
```
And another.

wjwwood commented:
+1 from code review

dirk-thomas commented:
@wjwwood I addressed all your comments in https://github.com/dirk-thomas/ros_comm/commit/52a7546fbed7c76f79af6bd5a0a4b9d6802e810f.

Superseded by #358.
